### PR TITLE
Feature/podpec validation

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

This PR introduces top-level validation for both legacy and v2 pod spec.

## QA steps

- deploy a CaaS charm with `unknown` fields in pod spec - now Juju will complain about the field;

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1851571

### CODE REVIEW

File: caas/specs/legacy.go [hunk 1]
```diff
@@ -11,24 +11,12 @@ import (
 // a pod on the CAAS substrate.
 type PodSpecLegacy struct {
 	podSpecBase `yaml:",inline"`
-
-	// legacy version has containers/initContainers two blocks.
-	InitContainers []ContainerSpec `json:"initContainers" yaml:"initContainers"`
```
Comment (ycliuhw) [inline caas/specs/legacy.go]: `InitContainers` and `Containers` are moved to the provider level. 
They are combined and filled into `Containers` with `Init bool` properly set.

**Comment by ycliuhw** [comment]
[pod spec yaml validation](https://trello.com/c/tlOmmfp1/975-pod-spec-yaml-validation)

**Comment by ycliuhw** [comment]
!!build!!
failed to fetch source for https://go.googlesource.com/oauth2: unable to get repository

File: caas/kubernetes/provider/specs/decode.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	"io/ioutil"
+	"strings"
+
+	"github.com/juju/errors"
+	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+// YAMLOrJSONDecoder attempts to decode a stream of JSON documents or
+// YAML documents by sniffing for a leading { character.
+type YAMLOrJSONDecoder struct {
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/decode.go]: can't we use the one from the k8s sdk?

Comment (ycliuhw) [reply caas/kubernetes/provider/specs/decode.go]: The reason why we have this new decoder is that they don't expose a way to enable strict mode for validation.
You can see it's still using the converter from the SDK to convert YAML to JSON then use golang JSON decoder to unmarshal the JSON string.
One thing we do here is to set `DisallowUnknownFields` before decoding.

**Comment by ycliuhw** [comment]
!!build!!
failed to fetch source for https://go.googlesource.com/oauth2: unable to get repository

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
failed to fetch source for https://go.googlesource.com/oauth2: unable to get repository
$$merge$$

**Pull request merged**